src/main/java/org/jenkinsci/plugins/workflow/support/steps/ExecutorStepExecution.java

```diff
@@ -416,6 +432,70 @@ private Object readResolve() {
             return getDisplayName();
         }
 
+        /** hash code of list of heads */
+        private transient int lastCheckedHashCode;
+        private transient String lastEnclosingLabel;
+        @Restricted(NoExternalUse.class) // for Jelly
+        public @CheckForNull String getEnclosingLabel() {
```

Comment:
In fact I expect the implementation to be instantaneous. There was a bug in an earlier iteration which I corrected here and also in https://github.com/jenkinsci/workflow-cps-plugin/issues/145.